### ISSUE TITLE
 allow to disable a cron-entry

### ISSUE THREAD

**Issue by woutervh** [reinout/z3c.recipe.usercrontab#1]
new feature-request:

it should be possible to add a cron-entry that is commented out, or to disabled an active cront-entry from a different builsdout-profile.

```
[mycronjob]
recipe = z3c.recipe.usercrontab
times = 0 12 * * *
command = echo hello
disabled = True     # default to False for backwards compat.
```

would generate:

```
# Generated by /opt/buildouts/... [mycronjob]
#0 * * * *       echo hello
# END /opt/buildouts/... [mycronjob]
```

**Comment by reinout** [comment]
What would be the purpose?

You _could_ just comment out the section in your buildout config and the recipe would remove it from your crontab.

Disabling from another buildout config file would be possible with something like:

```
parts -= mycronjob
```

**Comment by mauritsvanrees** [comment]
I would also just remove the part from another buildout config.

Older versions of buildout cannot handle a `parts +=` and `parts -=` in the same file, which could be a reason to add this, but it makes more sense to update your zc.buildout version then.

**Comment by woutervh** [comment]
some clarification

1) operational: we have projects that run multiple cronjobs every minute

```
# Generated by /opt/buildouts/... [cronjob1]
 * * * * *     command1
# END /opt/buildouts/... [cronjob1]

# Generated by /opt/buildouts/... [cronjob2]
 * * * * *     command2
# END /opt/buildouts/... [cronjob2]

# Generated by /opt/buildouts/... [cronjob3]
 * * * * *     command3
# END /opt/buildouts/... [cronjob3]
```

For this, our current deployment is:
    1) run the buildout
    2) quickly disable these cronjobs (sometimes you just have a few seconds left)
    3) validate deployment and configs
    4) activate cronjobs

kinda ugly to say the least

2) conceptual:  a buildout-deployment is a self-contained directory
   an "active" cronjob added by buildout breaks this idea.

   for daemons like supervisor or zope:  you deploy and then you start manually
   similarly some people prefer activating cronjobs by hand.

   Being able to deploy "disabled" cronjobs allows for this

**Comment by reinout** [comment]
Have you tried adding an commented-out entry? So something like:

```
[mycronjob]
recipe = z3c.recipe.usercrontab
times = 0 12 * * *
command = # echo hello
```

There _might_ be some check in the recipe that might need to be relaxed to allow the comment character, but wouldn't that be much much more simpler to implement?

**Comment by reinout** [comment]
**Yes, this is already possible**.

Actually, I wanted to add an example in the documentation and it turned out it was already in there. I've reverted it there, as the example is clearer without the comment :-) See https://github.com/reinout/z3c.recipe.usercrontab/commit/afa40c497956906ca2b30b97ea174c0670d8325f

**Comment by reinout** [comment]
So: just add a comment marker in front of the command and you're ready! No need to change the recipe.

**Comment by reinout** [comment]
I've closed it, I guess using the comment marker is more than enough?

**Comment by woutervh** [comment]
1)  commenting out the command  -> still an active cronjob

```
[mycronjob]
recipe = z3c.recipe.usercrontab
times = 0 12 * * *
command = # echo hello
```

this will generate

```
 0 12 * * * # echo hello
```

2)  commenting out the times-option -> still an active cronjob

```
[mycronjob]
recipe = z3c.recipe.usercrontab
times = #0 12 * * *
command =  echo hello
```

this will generate

```
 # 0 12 * * * echo hello
```

what don't like about this, is that you need to duplicate the time-option to deactivate it. Different buildout-profiles extending from each other can get out of sync.
( Imagine managing a whole series of cronjobs for 10 different 10 machines)

**Comment by mauritsvanrees** [comment]
With "still an active cronjob" you mean that it still gets added in the crontab, even though commented out? It does not do anything, right? Well, at least with a commented out 'times' line.

`times = #` is enough to comment it out. No need to add the full time, unless there is some validation error then, which we should fix.

`disabled = true` (or maybe nicer: `enabled = false`) sounds more like something that might be useful as a general buildout option. So: buildout goes through all the parts, and when enabled is false it skips that part. But still, that is what `parts -=` is for.

Oh, an alternative idea: add a buildout option `enable_cronjobs`. Default is true. When you set it to false, the cronjob recipe would do nothing. (Well, on uninstall it should probably still remove existing cronjobs.) I would like that on my development machine. Would that help for your use case? Or do you require control over individual cronjob sections?

**Comment by reinout** [comment]
From what I understand, @WouterVH _does_ want the various entries in the crontab.

In his case, when updating a site, first django migrations have to be run (for instance) and some other tasks have to be performed **before** it is OK to actually start running the cronjobs. Am I correct, @WouterVH ?

If is reasonably simple to write a small script that goes through the crontab to comment-out entries before the buildout run and enable them again afterwards. But for that the entries _do_ need to be available in the crontab.

So: the use case is valid.

And the solution to comment something out works for a single buildout, but @WouterVH is right that you need to duplicate the line (plus or minus a comment sign) in buildouts that override the part.

So: I finally understand why the option is useful :-) @WouterVH, go ahead and write a pull request. It should be relatively straightforward. Just look at the changes in my #3 pull request and make similar adjustments.

**Comment by woutervh** [comment]
thanks guys, I'll create a pull-request for this

**Comment by reinout** [comment]
Oopsie, I referred to #1 instead of #2 in a pull request, so this one accidentally got closed :-)
Reopening it.

**Comment by reinout** [comment]
I just released 1.4 which fixes this. Can you check if it works for you?
